### PR TITLE
Rate control (WIP) & ath9k

### PR DESCRIPTION
**Pull request by huaqiwen**

### CODE REVIEW

File: ath9k.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+nav_order: 2
+---
+
+# `ath9k`
+
+## Overview
+
+The kernel network driver used throughout the project is `ath9k`. This page intends to provide an overview of the underlying logic of this driver. 
+
+The source code of the `ath9k` driver can be found here at [Elixir Bootlin](https://elixir.bootlin.com/linux/latest/source/drivers/net/wireless/ath/ath9k).
```
Comment (OscBacon) [inline ath9k.md:12]: You could say "found at" instead of "found here at"

File: ath9k.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+## Transmission
+
+In this driver, there exists a linked list of empty packet nodes (`tx_buf`), where each node is a C `struct` containing space to record the information of one to-be-transmitted packet.
```
Comment (OscBacon) [inline ath9k.md:16]: "of one to-be-transmitted packet." 
maybe "of an outgoing packet"?

File: ath9k.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+The kernel network driver used throughout the project is `ath9k`. This page intends to provide an overview of the underlying logic of this driver. 
+
+The source code of the `ath9k` driver can be found here at [Elixir Bootlin](https://elixir.bootlin.com/linux/latest/source/drivers/net/wireless/ath/ath9k).
+
+## Transmission
+
+In this driver, there exists a linked list of empty packet nodes (`tx_buf`), where each node is a C `struct` containing space to record the information of one to-be-transmitted packet. 
+
+When `ath9k` receives a request to send a packet, it tries to obtain (pop) an empty node from the linked list buffer (`tx_buf`), fill in the packet information into the node, and then pass this node to a transmission routine by appending it to a transmission queue. After the transmission is finished, the node is emptied and gets added back to the buffer. Since the linked list buffer has a limited size (which can be obtained using the `ATH_TXBUF` macro) and each transmission temporarily pops one of its node, the linked list can be empty under high volumn of transmission demand. In such case, a packet is dropped thus cannot be transmitted.
```
Comment (OscBacon) [inline ath9k.md:18]: "limited size (which can be obtained using the `ATH_TXBUF` macro)"
"fixed size of ATH_TXBUF" maybe?

Comment (OscBacon) [inline ath9k.md:18]: "pops one of its node"
" pops one of its nodes"

Comment (OscBacon) [inline ath9k.md:18]: "volumn"

Comment (OscBacon) [inline ath9k.md:18]: "a packet is dropped thus cannot be transmitted."
"a packet is dropped and thus cannot be transmitted."

File: ath9k.md [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+This linked list is directly obtained from the private (`priv`) field of the hardware that the driver controls (of type `ieee80211_hw`), which is of type `ath_softc` and is usually named `sc`. Then `sc` along with other arguments are passed to buffer setup rountines (`ath_tx_setup_buffer`), which tries to obtain a buffer node using routine `ath_tx_get_buffer`. In particular, `sc` has a field `tx` of type `ath_tx` which contains buffer linked list called `txbuf`, and the routine `ath_tx_get_buffer` will try to pop a node from `txbuf` and return it to the calling routines.
```
Comment (OscBacon) [inline ath9k.md:20]: "rountines"
"routines"

Comment (OscBacon) [inline ath9k.md:20]: "contains buffer"
"contains a buffer"

**Review by OscBacon — CHANGES_REQUESTED** [review]

File: ath9k.md (hunk 4 above)
Comment (OscBacon) [inline ath9k.md:20]: Maybe remove "in particular", and split into 2 sentences, one for sc, one for get buffer. It sounds at first like they are linked, but I'm guessing you're just detailing the two things you talked about earlier?

File: index.md [hunk 5]
```diff
@@ -6,6 +6,9 @@ nav_order: 1
 # UCSMA Docs
 This site contains documentations for all UCSMA projects. The content in this page intends to provide a short overview of each projects, while the details can be found in the subpages in the navigation sidebar. 
 
+## Network driver (`ath9k`)
+The kernel network driver used throughout the project is `ath9k`. We directly modify this driver to implement some projects such as the rate control.
```
Comment (OscBacon) [inline index.md:10]: implement "projects"
"features" or "utilties" maybe?

File: rate_control.md [hunk 6]
```diff
@@ -1,5 +1,31 @@
...
-title: Rate Control
+title: Rate Control (WIP)
 ---
 
-# Rate Control
+# Rate Control (WIP)
+
+## Overview
+
+The rate control protocol is the key part of the ultimate goal of UCSMA, which attempts to maintain fairness among different nodes in an ad-hoc wireless network. There are two parts of the rate control protocol:
```
Comment (OscBacon) [inline rate_control.md:9]: maybe remove "of the ultimate goal"

Comment (OscBacon) [inline rate_control.md:9]: Maybe the goal of UCSMA should be its own paragraph at the top?

File: rate_control.md [hunk 7]
```diff
@@ -1,5 +1,31 @@
...
+
+1. buffer enqueue delay: a packet will need to wait a certain time interval (positively proportional to the current buffer size) before being enqueued into the transmission buffer; 
+2. contention window adjustment: we set lower CW for nodes with fuller buffer to allow their packets to be send our quicker.
```
Comment (OscBacon) [inline rate_control.md:12]: "we set lower CW for nodes with fuller buffer to allow their packets to be send our quicker."
"we set a lower CW for nodes with fuller buffers to allow their packets to be sent out quicker."

File: rate_control.md [hunk 8]
```diff
@@ -1,5 +1,31 @@
...
+
+
+
+
+
+
+
+
+
+### Main functions
```
Comment (OscBacon) [inline rate_control.md:25]: We should maybe consider agreeing on a format for these?

File: ath9k.md (hunk 3 above)
Comment (huaqiwen) [reply ath9k.md:18]: better to mention ATH_TXBUF is a macro

File: ath9k.md (hunk 4 above)
Comment (huaqiwen) [reply ath9k.md:20]: Those two things are connected, as `ath_tx_get_buffer` is accessing fields in `sc` (`tx.txbuf`).

File: rate_control.md (hunk 6 above)
Comment (huaqiwen) [reply rate_control.md:9]: Maybe add that later, but it's a needed context here

File: rate_control.md (hunk 8 above)
Comment (huaqiwen) [reply rate_control.md:25]: thats a part of the overview section

**Review by huaqiwen — COMMENTED** [review]
All matters resolved

**Pull request merged**